src/interface/state_descriptor.cpp

```diff
@@ -0,0 +1,347 @@
+// state descriptor DOES not have any of its function pointers set.
+std::shared_ptr<StateDescriptor> ResolvePackages(Packages_t &packages) {
+  using std::string;
+  auto state = std::make_shared<StateDescriptor>("parthenon::resolved_state");
+
+  // The workhorse data structure. Uses sets to cache which variables
+  // are of what type.
+  DependencyTracker var_tracker;
+  DependencyTracker swarm_tracker;
+
+  // TODO(JMM): The sparse stuff here will need to be revisited,
+  // and likely become much simpler once we have dense on block.
+
+  // Begin awful boilerplate
+  // ======================================================================
+  // Helper functions for adding vars
+  auto add_private_var = [&](const string &package, const string &var,
+                             const Metadata &metadata) {
+    state->AddField(package + "::" + var, metadata);
+  };
+  auto add_provides_var = [&](const string &package, const string &var,
+                              const Metadata &metadata) {
+    state->AddField(var, metadata);
+  };
+  auto add_overridable_var = [&](const string &var, Metadata &metadata) {
+    state->AddField(var, metadata);
+  };
+  // sparse
+  auto add_private_sparse = [&](const string &package, const string &var,
+                                const Metadata &metadata) {
+    const std::string name = package + "::" + var;
+    for (auto &m : packages[package]->AllSparseFields().at(var)) {
+      state->AddField(name, m.second);
+    }
+  };
+  auto add_provides_sparse = [&](const string &package, const string &var,
+                                 const Metadata &metadata) {
+    for (auto &m : packages[package]->AllSparseFields().at(var)) {
+      state->AddField(var, m.second);
+    }
+  };
+  // swarm
+  auto AddSwarm = [&](StateDescriptor *package, const string &swarm,
+                      const string &swarm_name, const Metadata &metadata) {
+    state->AddSwarm(swarm_name, metadata);
+    for (auto &p : package->AllSwarmValues(swarm)) {
+      auto &val_name = p.first;
+      auto &val_meta = p.second;
+      state->AddSwarmValue(val_name, swarm_name, val_meta);
+    }
+  };
+  auto add_private_swarm = [&](const string &package, const string &var,
+                               const Metadata &metadata) {
+    AddSwarm(packages[package].get(), var, package + "::" + var, metadata);
+  };
+  auto add_provides_swarm = [&](const string &package, const string &var,
+                                const Metadata &metadata) {
+    AddSwarm(packages[package].get(), var, var, metadata);
+  };
+  auto add_overridable_swarm = [&](const string &swarm, const Metadata &metadata) {
+    state->AddSwarm(swarm, metadata);
+    for (auto &pair : packages) {
+      auto &package = pair.second;
+      if (package->SwarmPresent(swarm)) {
+        for (auto &pair : package->AllSwarmValues(swarm)) {
+          state->AddSwarmValue(pair.first, swarm, pair.second);
+        }
+        return;
+      }
+    }
+  };
```

Comment:
Sure, this could be re-written as inheritence. Stylistically, I just prefer a functional programming style. I'll change it.

Yeah, I know your PR will clean a lot of this up. But I wanted to introduce the desired behaviour.

Reply:
Yeah, I was just letting you know that hopefully this will get less messy in the near term.

Reply:
That _might_ be more clear.  The lambdas are a little gross, but I guess it's not as clear to me that breaking this out into a new class will make it significantly better.  Certainly could.  One hesitation I'd have in suggesting you worry too much about this right now is that @AndrewGaspar's work on sparse will already clean up part of this, and then I'm not sure but some comments on an impending cleanup of `Swarm`/`SwarmVariables` might clean up that part.  Will the latter help here?  If so, then a lot of this boilerplate will go away anyway, even without adding a new class.  So I guess what I'm saying is that the best case is that this is cleaned up _and_ no new class is created.  Not sure what the prospects are for this, though.  Just a thought.

Reply:
It could simplify substantially, I agree. But I'm not sure the need for some flexibility will entirely disappear.